torch/_guards.py

```diff
@@ -419,3 +419,40 @@ def make_guard(self, fn, is_volatile=False) -> Guard:
 
     def is_nn_module(self) -> bool:
         return self.guard_source().is_nn_module()
+
+
+def detect_fake_mode(inputs: Any = None):
+    """
+    Attempts to "detect" what the current fake mode is.  If there is one ambiently
+    available from TracingContext, we preferentially use that.  Otherwise, we
+    heuristically detect the fake mode via the following sources, in order of
+    priority:
+
+        - Currently active fake mode on stack
+        - Fake mode associated with passed in tensors (inputs does not
+          have to be flattened)
+    """
+    from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+    from torch.utils._pytree import tree_flatten
+
+    context = TracingContext.get()
+    if context is not None:
+        fake_mode = context.fake_mode
+        if fake_mode is not None:
+            return fake_mode
+
+    from torch.utils._python_dispatch import _get_current_dispatch_mode_stack
+
+    for m in reversed(_get_current_dispatch_mode_stack()):
+        if isinstance(m, FakeTensorMode):
+            return m
+
+    fake_mode = None
+    flat_inputs, _ = tree_flatten(inputs)
```

Comment:
what does tree_flatten return with `inputs = None`?

Reply:
empty list

Reply:
ok not empty list, but it's still fine

```
>>> from torch.utils._pytree import tree_flatten
>>> tree_flatten(None)
([None], *)
```